src/fsharp/CheckComputationExpressions.fs

```diff
@@ -950,7 +950,8 @@ let TcComputationExpression cenv env overallTy tpenv (mWhole, interpExpr: Expr,
             error(Error(FSComp.SR.tcConstructIsAmbiguousInComputationExpression(), m))
 
         | SynExpr.ImplicitZero m -> 
-            if isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env m ad "Zero" builderTy) then error(Error(FSComp.SR.tcRequireBuilderMethod("Zero"), m))
+            if (not enableImplicitYield) && 
+               isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env m ad "Zero" builderTy) then error(Error(FSComp.SR.tcRequireBuilderMethod("Zero"), m))
             Some (translatedCtxt (mkSynCall "Zero" m []))
```

Comment:
Could you explain more - what happens to the invalid ` (mkSynCall "Zero" m [])` in the case where `enableImplicitYield` is true?  (invalid because the method may not exist).

Why doesn't `translatedCtxt` end up using this expression ?

Reply:
I see, in the case of implicit yields the holeFill is discarded here https://github.com/dotnet/fsharp/blob/ac4d5471c46fe1036a6a55214275f3f3f75daf93/src/fsharp/CheckComputationExpressions.fs#L1456

I need to think about this briefly

Reply:
So that holeFill is only discarded when the StatementOrYield thing fails to check as a statement.

So please add test cases for the cases where the last thing is a statement, e.g.

```
let x = 
    [ 1; 
      if true then 2
      printfn "hello" ]
```
In these cases we do need a Zero, and a good error message if it's missing.  I suspect it will generate an error message but a little later as it checks the expression returned by `mkSynCall` , which is actually ok tbh

Reply:
OK I see you have this test case here: https://github.com/dotnet/fsharp/pull/10556/files/ac4d5471c46fe1036a6a55214275f3f3f75daf93#diff-ac5756e06cf81b30d7554703b9cd9cd75c87e327fb985dd262e01696e57a0426R41

and indeed it returns a different error message.

Then this is all ok.  It would be good to add a comment along the lines 

```
// In some cases the node produced by `mkSynCall "Zero" m []` may be discarded in the case
// of implicit yields - for example "list { 1; 2 }" when each expression checks as an implicit yield.
// If it is not discarded, the syntax node will later be checked and the existence/non-existence of the Zero method
// will be checked/reported appropriately (though the error message won't mention computation expressions
// like our other error messages for missing methods).
```

Reply:
It sounds like you've answered your first question:  This seems like a weird gap in implementation - because yeah, we're using that passed-in `translatedCtxt` to deal with the implicitly yielded Zero being transformed out - but that check for the Zero method stops us before we get there.

With regards to the testing, I suspect the error message in-place should work comfortably enough:
> "The type 'ListBuilder' does not define the field, constructor or member 'Zero'."

With that said, is another test strictly needed?  Part of the testing above handles the case of a direct unit return as the final implicitly yielded element - both in the positive and negative cases.  Would the call to a unit-producing function like printfn change the resolution semantics?

Reply:
Oh, the joys of Github not updating comments while I'm writing and doing some double-checking on my work :)
I'll add said comment presently!